Create Poetry template for Python projects

sujuka99 opened the pull request:
Fixes #12

sujuka99 committed:
Laid out a python template.

sujuka99 committed:
Added some CI-related files

disrupted commented on templates/python/run_mypy.sh:
```diff
@@ -0,0 +1,9 @@
+# change directory to the project root directory.
+cd "$(dirname "$0")"
+
+mypy --pretty --package package-name --package tests --namespace-packages
```
simply `mypy --pretty {package_name} tests` would be enough here.

sujuka99 replied:
done

disrupted commented on templates/python/hooks/hook_generate-fetchers.sh:
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
```
please remove everything related to this hook (this file and in pre-commit-config)

disrupted commented on templates/python/.github/workflows/python-poetry-release.yaml:
```diff
@@ -0,0 +1,57 @@
+runs:
+  using: "composite"
+  steps:
+    - name: Set up Poetry ${{ inputs.poetry-version }}
```
you could replace all this with a call to our reusable workflow https://github.com/bakdata/ci-templates/tree/main/.github/workflows#calling-the-workflow-3

```yaml
jobs:
  release:
    uses: bakdata/ci-templates/.github/workflows/python-poetry-release.yaml@main
    with:
      ...
```

sujuka99 committed:
Removed unwanted files and packages.
Replaced workflow templates with a call to the bakdata reusable
poetry workflow.

sujuka99 committed:
Removed comments from pyproject.toml, shortened run_mypy.sh

sujuka99 commented:
I am unsure whether I have done what the issue calls for, let me know if and what I have missed, please.

disrupted commented:
> I am unsure whether I have done what the issue calls for, let me know if and what I have missed, please.

yep, overall it looks good. I will leave some comments

disrupted commented on templates/python/.github/workflows/python-poetry-release.yaml:
```diff
@@ -0,0 +1,27 @@
+name: Default reusable workflow
```
Please give it a more meaningful name, perhaps related to the Poetry release workflow you're calling. I would propose "Publish" or something similar

sujuka99 replied:
"Publish" sounds good, changed it to that.

disrupted commented on templates/python/.github/workflows/python-poetry-release.yaml:
```diff
@@ -0,0 +1,27 @@
+    with:
+      release-type: patch # (Required) See more values at: https://python-poetry.org/docs/cli/#version
+      ref: my-awesome-ref # (Optional) if not set the ${{ github.event.repository.default_branch }} will fill the value. In this case the changes will be pushed to my-awesome-ref
+      publish-to-test: false # (Optional) By default the packages are published to TestPyPI. In this case the packages are published to PyPI
```
Let's set this to `true` by default to avoid accidents. Please also change the secret to `${{ secrets.TEST_PYPI_TOKEN }}`

sujuka99 replied:
Done, I think.

disrupted replied:
You also need to update the comment. Or you can remove the comments and just link to the template workflow in the CI templates repo. Perhaps that's better because then we don't need to keep it up to date it in both places.

sujuka99 replied:
I am not sure if I understood correctly. I deleted almost all comments and added a single comment linking to the template. Is that what you meant?

disrupted replied:
I was referring to the comments behind each workflow parameter, especially
> By default the packages are published to TestPyPI. In this case the packages are published to PyPI

since you changed it to `true`

sujuka99 replied:
This I got, but the linking part I'm not sure about.

disrupted replied:
yes, looks good

disrupted commented on templates/python/.github/workflows/python-poetry-release.yaml:
```diff
@@ -0,0 +1,27 @@
+    uses: bakdata/ci-templates/.github/workflows/python-poetry-release.yaml@main
+    with:
+      release-type: patch # (Required) See more values at: https://python-poetry.org/docs/cli/#version
+      ref: my-awesome-ref # (Optional) if not set the ${{ github.event.repository.default_branch }} will fill the value. In this case the changes will be pushed to my-awesome-ref
```
I think you can remove this because you're setting a branch in the _on push_ trigger

sujuka99 replied:
done

disrupted commented on templates/python/.github/workflows/python-poetry-release.yaml:
```diff
@@ -0,0 +1,27 @@
+    branches: [main]
+
+jobs:
+  call-workflow-passing-data:
```
please change the name
and add `name: ...` attribute

sujuka99 replied:
Changed it. It's more specific than the default, but lmk if you think it doesn't fit. I just wrote the 3 steps that are taken.

disrupted replied:
_release_ would be a more fitting name imo

sujuka99 replied:
changed

disrupted commented on templates/python/.pre-commit-config.yaml:
```diff
@@ -0,0 +1,32 @@
+---
```
please add a CI workflow that runs pre-commit, similar to https://github.com/bakdata/streams-explorer/blob/7e657ddca4c7278affeeb31f74a7034f920c28b4/.github/workflows/ci.yml#L11

sujuka99 replied:
I think that's done.

disrupted commented on templates/python/.pre-commit-config.yaml:
```diff
@@ -0,0 +1,32 @@
+    rev: 5.10.1
+    hooks:
+      - id: isort
+        args: ["--settings", "backend/setup.cfg"]
```
I think you can remove the `backend/` prefix here (and everywhere else). This is only needed for projects where we have a backend and frontend in different subdirectories

sujuka99 replied:
Ah, of course. done

disrupted commented on templates/python/pyproject.toml:
```diff
@@ -0,0 +1,37 @@
+[tool.poetry.scripts]
+start = "main:start"
+
+[tool.poetry.urls]
+"Bug Tracker" = ""
```
you can remove this

sujuka99 replied:
done

disrupted reviewed (changes requested):
you can exclude the `hooks` and `licenses` folders. They are not necessary for the template

sujuka99 committed:
implementing feedback, small changes

sujuka99 committed:
Added pre-commit checks, small fixes.

disrupted commented on templates/python/.github/workflows/pre-commit.yaml:
```diff
@@ -0,0 +1,52 @@
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ["3.8"]
```
let's use something more recent as default
```suggestion
        python-version: ["3.10", "3.11"]
```

disrupted replied:
fyi it should always match what's specified in `pyproject.toml` as Python depedency

sujuka99 replied:
done

disrupted commented on templates/python/.github/workflows/pre-commit.yaml:
```diff
@@ -0,0 +1,52 @@
+
+jobs:
+  pre-commit:
+    name: pre-commit
```
name _test_

sujuka99 replied:
done

disrupted commented on templates/python/.github/workflows/pre-commit.yaml:
```diff
@@ -0,0 +1,52 @@
+    steps:
+      - uses: actions/checkout@v3
+
+      # - name: Install packages
+      #   run: sudo apt-get -y install python3-dev graphviz libgraphviz-dev pkg-config python3-wheel
```
remove

sujuka99 replied:
done

sujuka99 replied:
done

disrupted commented on templates/python/.github/workflows/pre-commit.yaml:
```diff
@@ -0,0 +1,52 @@
+      #   run: sudo apt-get -y install python3-dev graphviz libgraphviz-dev pkg-config python3-wheel
+
+      - name: Set up Poetry
+        uses: bakdata/ci-templates/actions/python-setup-poetry@v1.5.1
```
latest version of this action

```suggestion
        uses: bakdata/ci-templates/actions/python-setup-poetry@v1.5.3
```

sujuka99 replied:
done

disrupted commented on templates/python/.github/workflows/publish.yaml:
```diff
@@ -0,0 +1,29 @@
+
+on:
+  push:
+    branches: [master]
+  pull_request:
+    branches: [master]
```
new repositories use _main_ by default

sujuka99 replied:
True, for some reason I decided I had to use master, because this repo is with the old naming. Needless to say, not a bright moment. :D

disrupted replied:
no worries

sujuka99 committed:
small fixes

sujuka99 committed:
changed master to main

sujuka99 committed:
small fixes

sujuka99 committed:
changed name of job to release

disrupted commented on templates/python/.github/workflows/pre-commit.yaml:
```diff
@@ -0,0 +1,49 @@
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ["3.19", "3.11"]
```
typo in 3.10

disrupted commented on templates/python/pyproject.toml:
```diff
@@ -0,0 +1,34 @@
+pytest-cov = "^4.0.0"
+pytest-mock = "^3.6.1"
+
+[tool.poetry.scripts]
+start = "main:start"
```
I am unsure if we should leave this in. It's nice to have an entrypoint for applications where it makes sense (e.g. API server), but without defining the function in `main.py` it doesn't do anything

disrupted replied:
also not applicable to all application types. libraries would not have an entrypoint like this for example

sujuka99 replied:
I would suggest commenting it out without deletion, but I'm not sure if that's a good practice.

disrupted replied:
yeah, let's do that and wait for other opinions

sujuka99 replied:
done

sujuka99 committed:
typo

disrupted commented on templates/python/.github/workflows/publish.yaml:
```diff
@@ -0,0 +1,29 @@
+jobs:
+  release:
+    name: release
+    uses: bakdata/ci-templates/.github/workflows/python-poetry-release.yaml@main
```
should probably use a tag instead of main

sujuka99 replied:
true, used latest tag in `publish.yaml` and in `pre-commit.yaml`.

disrupted commented on templates/python/.github/workflows/publish.yaml:
```diff
@@ -0,0 +1,29 @@
+    name: release
+    uses: bakdata/ci-templates/.github/workflows/python-poetry-release.yaml@main
+    with:
+      release-type: patch publish-to-test: true
```
something went wrong here, those two parameters are joined

sujuka99 replied:
ah true, fixed it.

disrupted commented on templates/python/.github/workflows/publish.yaml:
```diff
@@ -0,0 +1,29 @@
+    with:
+      release-type: patch publish-to-test: true
+      python-version: 3.8
+      poetry-version: 1.1.11
```
unless there's a specific reason to overwrite it, I think we can just use the default (remove this line or comment out). A tagged version of this workflow (not `main`) will always use a specific version and won't change unexpectedly

sujuka99 replied:
commented it out

disrupted replied:
it still shows up for me? please double check that you commented out `poetry-version`

sujuka99 replied:
fixed

disrupted commented on templates/python/.github/workflows/publish.yaml:
```diff
@@ -0,0 +1,29 @@
+    uses: bakdata/ci-templates/.github/workflows/python-poetry-release.yaml@main
+    with:
+      release-type: patch publish-to-test: true
+      python-version: 3.8
```
this should also match what's specified in `pyproject.toml`
```suggestion
      python-version: "3.10"
```

the quotes could be important for 3.10

sujuka99 replied:
done

sujuka99 committed:
removed  entrypoint

sujuka99 committed:
updated values

sujuka99 committed:
commented out poetry version

disrupted commented on templates/python/.pre-commit-config.yaml:
```diff
@@ -0,0 +1,32 @@
+      - id: isort
+        args: ["--settings", "setup.cfg"]
+  - repo: https://github.com/psf/black
+    rev: 22.6.0
```
before we merge this, could you go through this list of tools and update each of them to the latest version? for example Black is already at 22.12.0, so we shouldn't use this old version for new projects.
Then the same version needs to be applied in `pyproject.toml`. I know it's a bit tedious but it would be great to get this in :)

sujuka99 replied:
On it

sujuka99 replied:
done

disrupted replied:
perfect, thanks!

disrupted commented on templates/python/.github/workflows/publish.yaml:
```diff
@@ -0,0 +1,30 @@
+name: Publish
+# Template adapted from https://github.com/bakdata/ci-templates/tree/main/.github/workflows#python-poetry-release
+
+on:
+  push:
+    branches: [main]
+  pull_request:
+    branches: [main]
```
I think we should change the default trigger to only publish from version tags

```suggestion
on:
  push:
    tags: ["v[0-9]+.[0-9]+.[0-9]+"]
```

sujuka99 replied:
Makes sense, yes. done

disrupted commented on templates/python/.github/workflows/pre-commit.yaml:
```diff
@@ -0,0 +1,49 @@
+
+on:
+  push:
+    # branches: [main]
+    tags: ["v[0-9]+.[0-9]+.[0-9]+"]
```
tests/checks should always run

```suggestion
    branches: [main]
    tags: ["v[0-9]+.[0-9]+.[0-9]+"]
```

sujuka99 replied:
Ah I have misunderstood how the tag branches works. I thought that if not specified the default value is the set of all branches.

sujuka99 replied:
uncommented

sujuka99 committed:
eof newlines added, updated dependencies versions
+ misc small tweaks

disrupted commented on templates/python/.github/workflows/publish.yaml:
```diff
@@ -0,0 +1,27 @@
+name: Publish
+# Template adapted from https://github.com/bakdata/ci-templates/tree/main/.github/workflows#python-poetry-release
+
+on:
+  push:
+    tags: ["v[0-9]+.[0-9]+.[0-9]+"]
```
oh, I actually made a mistake here. The `python-poetry-release` workflow does the version bumping and release.
Therefore we should call it using a manual workflow call like this, where we select the type of release (patch, minor, major) from a dropdown.

```suggestion
on:
  workflow_dispatch:
    inputs:
      release-type:
        description: "Scope of the release; see https://python-poetry.org/docs/cli/#version"
        type: choice
        required: true
        default: patch
        options:
          - patch
          - minor
          - major
```

then you can use the `${{ github.event.inputs.release-type }}` input for the `release-type` parameter below

see https://github.com/bakdata/kpops/blob/c4499753bac1f5f6a8c4b24a8f560d7cd66dcb4c/.github/workflows/release.yaml for reference

sujuka99 replied:
That's really nice! I updated the file. I didn't find anything I needed to change when using what you suggested, is that right?

disrupted replied:
yep, looks correct

sujuka99 committed:
updated workflow publish

disrupted commented on templates/python/.github/workflows/pre-commit.yaml:
```diff
@@ -0,0 +1,49 @@
+name: Pre-commit
```
I think I would prefer something more general such as _CI_ for the name of this workflow (and filename)
that's the same we already use for both kpops and streams-explorer

or do you have an idea?

sujuka99 replied:
_CI_ is good. I did notice _CI_ was used in the other projects, but I wasn't sure if it fits as the file looks a lot more specific than the _CI_'s you mentioned. For example in streams-explorer's ci.yml,  _Publish_ is also a job.

sujuka99 committed:
pre-commit.yaml is now called ci

disrupted commented on templates/python/.github/workflows/ci.yaml:
```diff
@@ -0,0 +1,49 @@
+jobs:
+  test:
+    name: test
+    runs-on: ubuntu-latest
```
should we use a fixed version here?
wdyt @yannick-roeder

yannick-roeder replied:
I would prefer that

sujuka99 replied:
Pinned to `ubuntu-22.04`. The latest _release_ image is `ubuntu-18.04`, but currently `ubuntu-latest` is equivalent to the _pre-release_ `ubuntu-22.04`.

yannick-roeder commented on templates/python/.github/workflows/ci.yaml:
```diff
@@ -0,0 +1,49 @@
+on:
+  push:
+    branches: [main]
+    tags: ["v[0-9]+.[0-9]+.[0-9]+"]
```
Do tags always start with a `v` in Python Poetry repos? Why not allow all tags here?

sujuka99 replied:
I couldn't answer tbh. Are you suggesting that we should allow all strings as in ["(.*)?"] ?

yannick-roeder replied:
@disrupted What do you think? You have more experience with Python projects?

disrupted replied:
it is certainly not a convention for Python projects that it should use semantic versioning with the `v` prefix. I believe for our projects it is the case though.
If we're sure that we will only use tags for official releases, then we can remove the condition there.

```suggestion
    tags:
```

yannick-roeder replied:
Sounds good :+1:

yannick-roeder commented on templates/python/.github/workflows/publish.yaml:
```diff
@@ -0,0 +1,37 @@
+
+  print-version:
+    name: print-version
+    runs-on: ubuntu-latest
```
Same here, please pin to the most recent version

sujuka99 replied:
done

yannick-roeder commented on templates/python/.github/workflows/publish.yaml:
```diff
@@ -0,0 +1,37 @@
+    runs-on: ubuntu-latest
+    needs: publish
+    steps:
+      - run: echo Bumped Version from ${{ needs.publish.outputs.old-tag }} to ${{ needs.publish.outputs.release-tag }}
```
It might be a good idea to test your workflows before merging this MR. I am not sure if this works correctly without quotes:
```suggestion
      - run: echo "Bumped Version from ${{ needs.publish.outputs.old-tag }} to ${{ needs.publish.outputs.release-tag }}"
```

sujuka99 replied:
Added the quotes, will test it in the playground

sujuka99 committed:
pinned workflow os

sujuka99 committed:
ci.yaml allows any tag

disrupted reviewed (approved):
LGTM

sujuka99 committed:
fixed errors exposed by testing
